### ISSUE TITLE
[porting] IAR's overload resolution struggles with `etl::span::span(etl::array&)`

### ISSUE THREAD

**Issue by dhebbeker** [ETLCPP/etl#753]
I try to construct a `etl::span` from an `etl::array`. This shouldn't be any issue as an appropriate constructor exists. 

I have [an example on Compiler Explorer][godbold] which compiles successfully. It uses

- avr gcc 13.2.0
- ETLCPP 20.38.0
- C++98

In short it does

```c++
etl::array<int,10> mArray;
etl::span<int> mSpan(mArray);
```

Now with the compiler I have to use I get this error:

```plain
Error[Pe309]: more than one instance of constructor "etl::span<T, 65535U>::span [with T=int]" matches the argument list:
            function template "etl::span<T, 65535U>::span(etl::array<U, N> &, etl::enable_if<etl::is_same<int, etl::remove_cv<U>::type>::value, void>::type *) [with T=int]" 
            function template "etl::span<T, 65535U>::span(TContainer &, etl::enable_if<<expression>, void>::type *) [with T=int]" 
            argument types are: (Array) 
```

- IAR C/C++ Compiler for AVR 6.70.1 (6.70.1.929)
- ETLCPP 20.38.1 (7fdaa7c9a461bfd191ee34ad950e01c919c8e7cd)
- Extended [Embedded C++](http://www.caravan.net/ec2plus/language.html) (a dialect of C++98)

The problem seems to be that IAR's overload resolution recognizes both constructor candidates as equally good during overload resolution.

Before suggesting a patch I wonder: Why are there both definitions:

- `span(etl::array<U, N>& a)` in [line 534](https://github.com/ETLCPP/etl/blob/f9da1cde9bace4edd9f1ef4edadf44daac9eb877/include/etl/span.h#L534) and
- `span(TContainer& a)` in [line 592](https://github.com/ETLCPP/etl/blob/f9da1cde9bace4edd9f1ef4edadf44daac9eb877/include/etl/span.h#L592C19-L592C37)

The latter does cover the first case equally well?

I think the code is currently correct as it is. But it contains (unnecessary?) overloads which the IAR chokes on.

[godbold]: https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAKwA7KSuADJ4DJgAcj4ARpjEEtIADqgKhE4MHt6%2BAcGp6Y4C4ZExLPGJUraY9kUMQgRMxATZPn5B1bWZDU0EJdFxCUm2jc2tuR0Ko30RA%2BVDUgCUtqhexMjsHOYAzBHI3lgA1CbbbpgEtAD0CsmGAHQIJ9gmGgCCO3sHmMen51dNxCYAE8Hk8Xu8zLsGPsvEcTmcLtdGo5kAB9JgKJTNUHbZ5vD7Qr4/BFXAhA5KYVEEQGEBQ4vHvN5kilYKiHP4gEAA4HwiIEUhcDRPQ6vYiAoEnKxM8mYVnsi6cm6GeGi8WcgBuYi8lOZmGFQluDEl4PBl0uh34xHltE53Ilb2wABVQqihI7Xo6AJJuVGvIRCbAAJUdEHBh3Dh1DbwjMetnLwClRClYetOYdjGY5IGImBYqHVlOQ6vhBsMGq1OplT05utI6YzMazObzBdRRZVYuB5e8lYp1ZAuv7mp7y3rxzMZizdvhwvQqEwCkODFQBAthEO8VEXiUhy8DCwxFoQIiwEOusXc6XK8OLCYBGQjwni2N%2BLeZotJEOjo8giYs2I4JOi6boet6vr%2BkGIZjlGrwNocYBgFmCaoqkfIJPCTa5vmlI5lQCSMBsHbiv2g64t22qjtGEbmGY36yH%2BkRWnOC5XquNCrpuTDbt864JocTCHKhggJGeVZPi%2BrxAa67pej6foBsGMExkpDYIUhiZCQQ6G/Aq2ZYa2uH4dCqZuKqPJkQOVYWcOFF1lR4Y0XRv7/oczGLsubHrpx3GHLxi4CZpImDuJ2xSpJzrSaBckQYp0FjjGam6ch06nGZErWRWlGwdRE5OY0Lluaxa4cZgW47n5/GHG4AC0UxAvQ/GdhKIVhVJIGyeBClQfZkbxRG6lJimM49XBmEtoWxanKWfggDZvaptgNYynZ2VweGY3YW2k0kpyzabYZObGUR5mLZZFLkfNJFWadc1ZTGjk/vljGufO7nXuxG6lVx5UEIue4HkeJ6iRSL0sR5N53g%2BNHPqFJpvGlN5pRJ003tNEAsGlMNheCfIfqgECLMcgRheGOYEGsDCo4adzpAAXpgBMSSYgQACJwxCuxsghrKzOgECvAAaoGhPmAAbGL8FgDzkR86ivpC8%2B%2BKQp8sLfBhiIFsQ6QCPSJrK4SqvEk4Uw5qwYJK1CMJwqcyBTOgtB4LE5uMq8uO3hEBNEyTEbbkDzBsEqGyHHbEkxmgXirjOpzjmYihA0Bhya9rlM0fdabR1JAtBkInoAPJRKiACyrwAFK54GxLwjHdw0X1a31w3UduIcmfZ3nBeF56UTl5X0c0TXE51w3w8Zk3LcRVngY5/nqLKB6bgABK983ri0KHEZkxTLcABqeo6rpyG4bjYP6TOs3r2yuHgVAcMsNocP4vB%2BBwWikKgnBuNY1jB6s6xq5CPBSAEE0LfZYABrEAkhJB3ECBofwgpAj%2BAABz%2BH8AATn8GYaQ99JBPxAW/TgvAFAgA0EAkByw4CwBgIgEAqwCDJAjuQSgaAWDJDoAkKIKZOCqCQaLaqotJCHGAMgZAhwuDbDuGYXgspCAkDwOgPQ/BBAiDEOwKoSj5BKDUPg3QXBSAAHdATJE4DwO%2BnBH6kGfq/d%2BHBc4RwYauVAbJBYV2EaI8Rki7gaEjB4Vh7CrQ7C4IsXgwCX6gNIAgTATADyUDMRwXBpAWABFIVY3gNiiEkLIWEih1CIBIBYWw%2BgZAKAQAKf4rk6piDAHEWYUhWB1R4A2ALPAmB9G5wpM/QBNBaBaS1pQWI%2BDYgRCaECExvAhnMGIECXOsRtClVCYAlhbBBC5wYEeMZpAsCxC8MANwYhaDEO4LwLAt4jDiDCZsvAOYHB4ALIc1%2BmBVClQjpsV%2BaF76vwdrEQEUyPBYA2dSPASSjnLCoAYYAChmmtPaYwDZGiVHiHUbIRQKh1AXN0foQwxgv6WH0I7YhsB/bsD%2BKQTWA5iB7jAYsZYqBkh1EObVAg6ATgs1MJYawaCkGHGqoYdI3LVCHDZVYSw3KWZSW7rPQMucABinpQjYFRAvbArwxUV2qmKiKEq3ShFFeK3OqI3DKGUKiKIrwC5CDkEa8ujpuWnmqrnaxmtiDyIXPAZYdh5mZBcPucYfg9FhFmGUCoegCgZAEL6kNaQw0MH6EGhYnRPUCB6GMTwbQ9AepuUm6YsbBiVBGL0CNeipi9BzfMSo7rf4bAkHEixqSCEcBFELIRIixESLMF4yMuBZEBIAcErJWhqWkAgagyRaDJBILQVgqQkgNCSECKLfQnAElJP8Ck/B6TbCZNCQO8BAR/CjvHZOyQ07Z3zsXRwbYeCLkbu3eE%2B%2BUjLHrsIf28JScvWSCAA

**Comment by jwellbelove** [comment]
I have a fix that I will push very soon.

**Comment by dhebbeker** [comment]
Thank you!

**Comment by jwellbelove** [comment]
Actually, I may have been too hasty. I've been working on #754, not this one.
What was your suggested fix?

**Comment by jwellbelove** [comment]
You are correct, `span(TContainer& a)` works fine for `std` & `etl` arrays.

**Comment by dhebbeker** [comment]
I would have removed `span(etl::array<U, N>& a)`. But I see you did this already in 01d96609cf5193f9a5b8f90cd8a1a23a5ea05631 👍

I tested that commit with the IAR compiler mentioned above. The code (same as the one referenced in the description) now compiles without compiler errors 🥳 

🙏 Thank you!

**Comment by jwellbelove** [comment]
Fixed 20.38.2
